src/core/thread/mlr_manager.cpp

```diff
@@ -0,0 +1,406 @@
+            addr->GetAddress().IsMulticastLargerThanRealmLocal() &&
+            MatchesFilter(*addr, kMulticastAddressFilterNotRegistered))
+        {
+            SuccessOrExit(error = message->Append(&addr->GetAddress(), sizeof(Ip6::Address)));
+            addr->SetMlrState(kMlrStateRegistering);
+
+            if (--num == 0)
+            {
+                break;
+            }
+        }
+    }
+
+    if (!mle.IsFullThreadDevice() && mle.GetParent().IsThreadVersion1p1())
+    {
+        uint8_t pbbrServiceId;
+
+        SuccessOrExit(error = Get<BackboneRouter::Leader>().GetServiceId(pbbrServiceId));
+        SuccessOrExit(error = mle.GetServiceAloc(pbbrServiceId, messageInfo.GetPeerAddr()));
+    }
+    else
+    {
+        messageInfo.GetPeerAddr().SetToRoutingLocator(mle.GetMeshLocalPrefix(),
+                                                      Get<BackboneRouter::Leader>().GetServer16());
+    }
+
+    messageInfo.SetPeerPort(kCoapUdpPort);
+    messageInfo.SetSockAddr(mle.GetMeshLocal16());
+
+    SuccessOrExit(error = Get<Coap::Coap>().SendMessage(
+                      *message, messageInfo, &MlrManager::HandleMulticastListenerRegistrationResponse, this));
+
+exit:
+    if (error == OT_ERROR_NONE)
+    {
+        mMlrPending = true;
+    }
+    else
+    {
+        if (message != nullptr)
+        {
+            message->Free();
+        }
+
+        SetNetifMulticastAddressMlrState(kMulticastAddressFilterRegistering, kMlrStateToRegister);
+
+        if (error == OT_ERROR_NO_BUFS)
+        {
+            ScheduleSend(1);
+        }
+    }
+
+    otLogInfoBbr("Send MLR.req: %s", otThreadErrorToString(error));
+    LogMulticastAddresses();
+}
+
+void MlrManager::HandleMulticastListenerRegistrationResponse(Coap::Message *         aMessage,
+                                                             const Ip6::MessageInfo *aMessageInfo,
+                                                             otError                 aResult)
+{
+    uint8_t status = ThreadStatusTlv::MlrStatus::kMlrGeneralFailure;
+    otError error;
+
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    mMlrPending = false;
+
+    VerifyOrExit(aResult == OT_ERROR_NONE && aMessage != nullptr, error = OT_ERROR_PARSE);
+    VerifyOrExit(aMessage->GetCode() == OT_COAP_CODE_CHANGED, error = OT_ERROR_PARSE);
+
+    SuccessOrExit(error = Tlv::FindUint8Tlv(*aMessage, ThreadTlv::kStatus, status));
+
+exit:
+
+    if (status == ThreadStatusTlv::MlrStatus::kMlrSuccess)
+    {
+        SetNetifMulticastAddressMlrState(kMulticastAddressFilterRegistering, kMlrStateRegistered);
+        // keep sending until all multicast addresses are registered.
+        ScheduleSend(0);
+    }
+    else
+    {
+        otBackboneRouterConfig config;
+        uint16_t               reregDelay;
+
+        SetNetifMulticastAddressMlrState(kMulticastAddressFilterRegistering, kMlrStateToRegister);
+
+        // Schedule reregistration if failed in t < Reregistration Delay
```

Comment:
Thread 1.2 Specification 5.24.2.3 should be the source for the test requirement.
update the comment?